diff --git a/CMakeLists.txt b/CMakeLists.txt
index 35b48770..7955fe73 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -23,10 +23,12 @@
 # LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE
 # OF THIS SOFTWARE.
 
-cmake_minimum_required(VERSION 2.8.11) # b/c of use of BUILD_INTERFACE generator expression
+# [cuCIM patch] Set minimum CMake version to 3.30.0
+cmake_minimum_required(VERSION 3.30.0)
 
-# Default policy is from 2.8.9
-cmake_policy(VERSION 2.8.9)
+# [cuCIM patch] Set default policy to 3.30.0
+# Default policy is from 3.30.0
+cmake_policy(VERSION 3.30.0)
 # Set MacOSX @rpath usage globally.
 if (POLICY CMP0020)
   cmake_policy(SET CMP0020 NEW)
