diff --git a/CMakeLists.txt b/CMakeLists.txt
index 35b48770..416377e7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -23,24 +23,8 @@
 # LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE
 # OF THIS SOFTWARE.
 
-cmake_minimum_required(VERSION 2.8.11) # b/c of use of BUILD_INTERFACE generator expression
-
-# Default policy is from 2.8.9
-cmake_policy(VERSION 2.8.9)
-# Set MacOSX @rpath usage globally.
-if (POLICY CMP0020)
-  cmake_policy(SET CMP0020 NEW)
-endif(POLICY CMP0020)
-if (POLICY CMP0042)
-  cmake_policy(SET CMP0042 NEW)
-endif(POLICY CMP0042)
-# Use new variable expansion policy.
-if (POLICY CMP0053)
-  cmake_policy(SET CMP0053 NEW)
-endif(POLICY CMP0053)
-if (POLICY CMP0054)
-  cmake_policy(SET CMP0054 NEW)
-endif(POLICY CMP0054)
+# [cuCIM patch] Set minimum CMake version to 3.30.0
+cmake_minimum_required(VERSION 3.30.0)
 
 # Read version information from configure.ac.
 FILE(READ "${CMAKE_CURRENT_SOURCE_DIR}/configure.ac" configure)
