diff --git a/CMakeLists.txt b/CMakeLists.txt
index b04561f4..2392c14d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -7,7 +7,8 @@
 # For this purpose you can define a CMake var: OPENJPEG_NAMESPACE to whatever you like
 # e.g.:
 # set(OPENJPEG_NAMESPACE "GDCMOPENJPEG")
-cmake_minimum_required(VERSION 3.5)
+# [cuCIM patch] Set minimum CMake version to 3.30.0
+cmake_minimum_required(VERSION 3.30.0)
 
 if(NOT OPENJPEG_NAMESPACE)
   set(OPENJPEG_NAMESPACE "OPENJPEG")
